File: isilibrary/src/main/java/com/isi/isilibrary/products/ManageElementsActivity.java
package com.isi.isilibrary.products;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.SearchView;
import android.widget.Spinner;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.isi.isiapi.classes.Category;
import com.isi.isiapi.classes.CategoryAndProduct;
import com.isi.isiapi.classes.Product;
import com.isi.isilibrary.IsiAppActivity;
import com.isi.isilibrary.R;
import com.isi.isilibrary.backActivity.BackActivity;
import com.isi.isilibrary.products.recycler.ElementRecycler;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class ManageElementsActivity extends BackActivity {

    private CategoryAndProduct categorySelected = null;
    private RecyclerView layout;
    private final List<Product> products = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_manage_elements);

        setTitle("Gestisci elementi");

        layout = findViewById(R.id.product_list_recycler);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setOrientation(RecyclerView.VERTICAL);
        layout.setLayoutManager(linearLayoutManager);
    }

    @Override
    protected void onResume() {
        super.onResume();

        new Thread(() -> {

            products.clear();

            List<CategoryAndProduct> categoryAndProducts = IsiAppActivity.isiCashierRequest.getCategories();

            for (CategoryAndProduct categoryAndProduct : categoryAndProducts){
                products.addAll(categoryAndProduct.product);
            }

            products.sort(Comparator.comparing(one -> one.name));

            runOnUiThread(() -> {

                updateUi("");

                SearchView search = findViewById(R.id.searchElement);

                search.setClickable(true);
                search.setQueryHint("Cerca");

                search.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
                    @Override
                    public boolean onQueryTextSubmit(String s) {
                        return false;
                    }

                    @Override
                    public boolean onQueryTextChange(String s) {

                        final String s2 = s;

                        runOnUiThread(() -> updateUi(s2));

                        return true;
                    }
                });

            });


        }).start();

    }

    private void updateUi(final String s){

        ElementRecycler recycler;
        if(categorySelected != null) {
            if (categorySelected.category.id == 0) {
                recycler = new ElementRecycler(this, products.stream()
                        .filter(prod -> prod.name.toLowerCase().contains(s.toLowerCase()))
                        .collect(Collectors.toList()));
            } else {
                recycler = new ElementRecycler(this, products.stream()
                        .filter(prod -> prod.name.toLowerCase().contains(s.toLowerCase()) && prod.category_id == categorySelected.category.id)
                        .collect(Collectors.toList()));
            }

            layout.setAdapter(recycler);
        }



    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.manage_element_menu, menu);

        MenuItem item = menu.findItem(R.id.manage_element_spinner);
        Spinner spinner = (Spinner) item.getActionView();

        new Thread(() -> {
            List<CategoryAndProduct> categories = IsiAppActivity.isiCashierRequest.getCategories();
            CategoryAndProduct cat = new CategoryAndProduct();
            cat.category = new Category(0, "Tutto", 0, "");
            categories.add(0, cat);

            runOnUiThread(() -> {
                ArrayAdapter<CategoryAndProduct> adapter = new ArrayAdapter<>(ManageElementsActivity.this,android.R.layout.simple_spinner_item, categories);
                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

                spinner.setAdapter(adapter);

                spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                        if(categories.get(position) != categorySelected){
                            categorySelected = categories.get(position);
                            updateUi("");
                        }
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> parent) {

                    }
                });
            });


        }).start();

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if(item.getItemId() == R.id.addIntestazioneMenu){
            Intent intent = new Intent(ManageElementsActivity.this, AddManageElementActivity.class);
            startActivity(intent);
        }

        return super.onOptionsItemSelected(item);
    }
}

File: isilibrary/src/main/java/com/isi/isilibrary/products/ManageElementsActivity.kt
package com.isi.isilibrary.products

import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.*
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.isi.isiapi.classes.*
import com.isi.isiapi.classes.isiorder.CategoryAndListini
import com.isi.isilibrary.IsiAppActivity
import com.isi.isilibrary.R
import com.isi.isilibrary.backActivity.BackActivity
import com.isi.isilibrary.dialog.NetConnection
import com.isi.isilibrary.products.recycler.ElementRecycler
import kotlin.collections.ArrayList

class ManageElementsActivity : BackActivity() {
    private lateinit var categorySelected: Category
    private lateinit var layout: RecyclerView
    private lateinit var recycler: ElementRecycler
    private var products: MutableList<Product> = ArrayList()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_manage_elements)
        title = "Gestisci elementi"

        val cat = Category(0, "Tutto", 0, "", 0)
        categorySelected = cat

        layout = findViewById(R.id.product_list_recycler)
        val linearLayoutManager = LinearLayoutManager(this)
        linearLayoutManager.orientation = RecyclerView.VERTICAL
        layout.layoutManager = linearLayoutManager
    }

    override fun onResume() {
        super.onResume()

        NetConnection<CategoryAndListini>(this,
                "Scarico elementi...",
                startNetConnection = {
                    IsiAppActivity.httpRequest!!.categories
                },
                onConnectionOk = {
                    products.clear()

                    for (categoryAndProduct in it.categories) {
                        products.addAll(categoryAndProduct.products!!)
                    }

                    products.sortBy { it2 -> it2.name.lowercase() }

                    recycler = ElementRecycler(this, products)
                    layout.adapter = recycler
                    recycler.search(categorySelected.id, "")
                    val search = findViewById<SearchView>(R.id.searchElement)
                    search.isClickable = true
                    search.queryHint = "Cerca"
                    search.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                        override fun onQueryTextSubmit(s: String): Boolean {
                            return false
                        }

                        override fun onQueryTextChange(s: String): Boolean {
                            recycler.search(categorySelected.id, s)
                            return true
                        }
                    })
                },
                onConnectionError = {

                }
        ).start()

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.manage_element_menu, menu)
        val item = menu.findItem(R.id.manage_element_spinner)
        val spinner = item.actionView as Spinner?
        Thread {
            val categories: CategoryAndListini? =
                    IsiAppActivity.httpRequest!!.categories

            if (categories != null) {
                val cat = Category(0, "Tutto", 0, "", 0)
                categories.categories.add(0, cat)
                categorySelected = cat
                runOnUiThread {
                    val adapter = ArrayAdapter(
                            this@ManageElementsActivity,
                            android.R.layout.simple_spinner_item,
                            categories.categories
                    )
                    adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                    spinner!!.adapter = adapter
                    spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                        override fun onItemSelected(
                                parent: AdapterView<*>?,
                                view: View,
                                position: Int,
                                id: Long
                        ) {
                            if (categories.categories[position] !== categorySelected) {
                                categorySelected = categories.categories[position]
                                recycler.search(categorySelected.id, "")
                            }
                        }

                        override fun onNothingSelected(parent: AdapterView<*>?) {}
                    }
                }
            }

        }.start()
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.addIntestazioneMenu) {
            val intent = Intent(this@ManageElementsActivity, AddManageElementActivity::class.java)
            startActivity(intent)
        }
        return super.onOptionsItemSelected(item)
    }
}